libfastpm/pm2lpt.c

```diff
@@ -50,21 +58,32 @@ pm_2lpt_solve(PM * pm, FastPMFloat * delta_k, FastPMStore * p, double shift[3],
     }
     FastPMFieldDescr DX1[] = { {COLUMN_DX1, 0}, {COLUMN_DX1, 1}, {COLUMN_DX1, 2}};
     FastPMFieldDescr DX2[] = { {COLUMN_DX2, 0}, {COLUMN_DX2, 1}, {COLUMN_DX2, 2}};
+    FastPMFieldDescr DV1[] = { {COLUMN_DV1, 0}, {COLUMN_DV1, 1}, {COLUMN_DV1, 2}};  //This will do nothing throughout the function if !p->v1
     int D1[] = {1, 2, 0};
     int D2[] = {2, 0, 1};
 
+    /* 1LPT */
     for(d = 0; d < 3; d++) {
 
         fastpm_apply_laplace_transfer(pm, delta_k, workspace, potorder);
         fastpm_apply_diff_transfer(pm, workspace, workspace, d);
 
+        if (p->dv1) {
```

Comment:
Is there a way to avoid workspace_v1?
Always allocating workspace_v1 drastically increase memory usage. (by 20% as a quick estimate)

One posibility is to do an extra loop if (p->dv1) and redo some of the transfer functions. You don't have to worry about overwriting conflicts with the original DX1 calculation.